File: src/main/java/com/smarthome/backend/server/api/modules/hue/HueBridgeController.java
package com.smarthome.backend.server.api.modules.hue;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.security.cert.X509Certificate;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;

import javax.net.ssl.HttpsURLConnection;
import javax.net.ssl.SSLContext;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.smarthome.backend.model.Module;
import com.smarthome.backend.server.actions.ActionManager;
import com.smarthome.backend.server.api.ApiRouter;
import com.smarthome.backend.server.api.modules.ModuleEventStreamManager;
import com.smarthome.backend.server.db.DatabaseManager;
import com.smarthome.backend.server.db.JsonRepository;
import com.smarthome.backend.server.db.Repository;

/**
 * Kapselt die Logik zum Pairing mit einer Hue Bridge über die Hue CLIP API.
 */
public class HueBridgeController {

    private static final Logger logger = LoggerFactory.getLogger(HueBridgeController.class);
    private static final Gson gson = new Gson();
    
    private final Repository<HueDiscoveredBridge> bridgeRepository;
    private final Repository<Module> moduleRepository;
    private final DatabaseManager databaseManager;
    
    public HueBridgeController() {
        this.bridgeRepository = null;
        this.moduleRepository = null;
        this.databaseManager = null;
    }
    
    public HueBridgeController(DatabaseManager databaseManager) {
        this.databaseManager = databaseManager;
        if (databaseManager != null) {
            this.bridgeRepository = new JsonRepository<>(databaseManager, HueDiscoveredBridge.class);
            this.moduleRepository = new JsonRepository<>(databaseManager, Module.class);
        } else {
            this.bridgeRepository = null;
            this.moduleRepository = null;
        }
    }


    /**
     * Stellt über die Hue CLIP API ein Pairing zur angegebenen Bridge her.
     * Die Bridge wird anhand der bridgeId aus der Datenbank geladen.
     * 
     * Gemäß Hue API V2 Getting Started Guide:
     * - POST Request an http://<bridge ip address>/api
     * - Body: {"devicetype":"app_name#instance_name", "generateclientkey":true}
     * - Vor dem Aufruf muss der physische Link-Button an der Bridge gedrückt werden
     * 
     * @param exchange HTTP-Exchange für die Antwort
     * @param bridgeId Die deviceId der Bridge (aus dem Discovery-Suchlauf)
     * @throws IOException bei Fehlern beim Senden der Antwort
     */
    public void pairBridge(com.sun.net.httpserver.HttpExchange exchange, String bridgeId, Runnable onPairingSuccess) throws IOException {
        try {
            if (bridgeRepository == null) {
                ApiRouter.sendResponse(exchange, 500,
                    gson.toJson(Map.of("error", "DatabaseManager nicht initialisiert - Bridge kann nicht aus Datenbank geladen werden")));
                return;
            }
            
            logger.info("Starte Hue-Pairing für Bridge: {}", bridgeId);
            
            // Lade Bridge aus der Datenbank
            Optional<HueDiscoveredBridge> bridgeOpt = bridgeRepository.findById(bridgeId);
            if (!bridgeOpt.isPresent()) {
                ApiRouter.sendResponse(exchange, 404,
                    gson.toJson(Map.of("error", "Bridge mit ID '" + bridgeId + "' nicht in Datenbank gefunden. Bitte zuerst Discovery durchführen.")));
                return;
            }
            
            HueDiscoveredBridge bridge = bridgeOpt.get();
            String bridgeIp = bridge.getBestConnectionAddress();
            int port = bridge.getPort();
            
            if (bridgeIp == null || bridgeIp.isEmpty()) {
                ApiRouter.sendResponse(exchange, 400,
                    gson.toJson(Map.of("error", "Keine gültige IP-Adresse für Bridge '" + bridgeId + "' gefunden")));
                return;
            }
            
            logger.info("Bridge gefunden: {} -> {}:{}", bridgeId, bridgeIp, port);
            
            // Führe Pairing durch (POST https://{bridgeIp}:{port}/api)
            String apiUrl = "https://" + bridgeIp + ":" + port + "/api";
            URI uri = new URI(apiUrl);
            
            // Erstelle SSLContext, der alle Zertifikate akzeptiert
            TrustManager[] trustAllCerts = new TrustManager[] {
                new X509TrustManager() {
                    @Override
                    public X509Certificate[] getAcceptedIssuers() {
                        return new X509Certificate[0];
                    }
                    
                    @Override
                    public void checkClientTrusted(X509Certificate[] certs, String authType) {
                        // Akzeptiere alle Client-Zertifikate
                    }
                    
                    @Override
                    public void checkServerTrusted(X509Certificate[] certs, String authType) {
                        // Akzeptiere alle Server-Zertifikate
                    }
                }
            };
            
            SSLContext sslContext = SSLContext.getInstance("TLS");
            sslContext.init(null, trustAllCerts, new java.security.SecureRandom());
            
            // Erstelle HTTPS-Verbindung mit dem Trust-All SSLContext
            HttpsURLConnection connection = (HttpsURLConnection) uri.toURL().openConnection();
            connection.setSSLSocketFactory(sslContext.getSocketFactory());
            connection.setHostnameVerifier((hostname, session) -> true); // Akzeptiere alle Hostnamen
            connection.setRequestMethod("POST");
            connection.setConnectTimeout(10000); // 10 Sekunden Timeout für Pairing
            connection.setReadTimeout(30000);
            connection.setDoOutput(true);
            connection.setRequestProperty("Content-Type", "application/json");
            
            // Erstelle Request-Body gemäß Hue API (Standard-Werte)
            JsonObject payload = new JsonObject();
            payload.addProperty("devicetype", "smarthome-backend#server");
            payload.addProperty("generateclientkey", true);
            
            String payloadString = payload.toString();
            logger.debug("Pairing-Request an {}: {}", apiUrl, payloadString);
            
            // Sende Request
            try (OutputStream os = connection.getOutputStream()) {
                byte[] input = payloadString.getBytes(StandardCharsets.UTF_8);
                os.write(input);
            }
            
            int status = connection.getResponseCode();
            InputStream is = status >= 200 && status < 300
                ? connection.getInputStream()
                : connection.getErrorStream();
            
            StringBuilder responseBuilder = new StringBuilder();
            try (BufferedReader reader = new BufferedReader(
                    new InputStreamReader(is, StandardCharsets.UTF_8))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    responseBuilder.append(line);
                }
            }
            
            String responseBody = responseBuilder.toString();
            logger.debug("Pairing-Antwort von Bridge {}: HTTP {} {}", bridgeIp, status, responseBody);
            
            // Parse Response (Array mit success/error-Objekten)
            JsonArray responseArray;
            try {
                JsonElement jsonElement = gson.fromJson(responseBody, JsonElement.class);
                if (jsonElement.isJsonArray()) {
                    responseArray = jsonElement.getAsJsonArray();
                } else {
                    ApiRouter.sendResponse(exchange, 500,
                        gson.toJson(Map.of("error", "Ungültige Antwort-Format: Erwartet JSON-Array")));
                    return;
                }
            } catch (Exception parseEx) {
                ApiRouter.sendResponse(exchange, 500,
                    gson.toJson(Map.of("error", "Fehler beim Parsen der Bridge-Antwort: " + parseEx.getMessage() + " (Response: " + responseBody + ")")));
                return;
            }
            
            // Prüfe Response auf Success/Error
            PairingResult result = new PairingResult();
            result.setHttpStatus(status);
            result.setRawResponse(responseBody);
            
            for (JsonElement element : responseArray) {
                if (element.isJsonObject()) {
                    JsonObject obj = element.getAsJsonObject();
                    
                    // Prüfe auf Success
                    if (obj.has("success")) {
                        JsonObject successObj = obj.getAsJsonObject("success");
                        String username = null;
                        String clientKey = null;
                        
                        if (successObj.has("username")) {
                            username = successObj.get("username").getAsString();
                            result.setSuccess(true);
                            result.setUsername(username);
                            logger.info("Pairing erfolgreich: Bridge {} -> Username: {}", bridgeId, username);
                        }
                        if (successObj.has("clientkey")) {
                            clientKey = successObj.get("clientkey").getAsString();
                            result.setClientKey(clientKey);
                            logger.info("Client-Key erhalten: {}", clientKey.substring(0, Math.min(10, clientKey.length())) + "...");
                        }
                        
                        // Aktualisiere Bridge mit Pairing-Informationen (beide Werte)
                        if (username != null) {
                            bridge.setIsPaired(true);
                            bridge.setUsername(username);
                            if (clientKey != null) {
                                bridge.setClientKey(clientKey);
                            }
                            
                            // Speichere aktualisierte Bridge in der Datenbank
                            try {
                                bridgeRepository.save(bridgeId, bridge);
                                logger.info("Bridge {} erfolgreich als gepaart gespeichert (username: {}, clientKey: {})", 
                                    bridgeId, username, clientKey != null ? clientKey.substring(0, Math.min(10, clientKey.length())) + "..." : "null");
                                
                                // Aktualisiere Bridge auch im Hue-Modul
                                updateBridgeInModule(bridgeId, bridge);
                                
                            } catch (Exception saveEx) {
                                logger.error("Fehler beim Speichern der Pairing-Informationen für Bridge {}", bridgeId, saveEx);
                                // Fehler beim Speichern ist nicht kritisch, Pairing war erfolgreich
                            }
                        }
                    }
                    
                    // Prüfe auf Error
                    if (obj.has("error")) {
                        JsonObject errorObj = obj.getAsJsonObject("error");
                        int errorType = errorObj.has("type") ? errorObj.get("type").getAsInt() : -1;
                        String errorDescription = errorObj.has("description") 
                            ? errorObj.get("description").getAsString() 
                            : "Unbekannter Fehler";
                        
                        result.setError(true);
                        result.setErrorType(errorType);
                        result.setErrorDescription(errorDescription);
                        
                        logger.warn("Pairing-Fehler: Bridge {} -> Type: {}, Description: {}", 
                            bridgeId, errorType, errorDescription);
                    }
                }
            }

            
            // Erstelle Response
            Map<String, Object> response = new java.util.HashMap<>();
            response.put("success", result.isSuccess());
            if (result.isSuccess()) {
                response.put("username", result.getUsername());
                if (result.getClientKey() != null) {
                    response.put("clientKey", result.getClientKey());
                }
                onPairingSuccess.run();
            } else {
                response.put("error", result.getErrorDescription());
                response.put("errorType", result.getErrorType());
            }
            response.put("httpStatus", result.getHttpStatus());
            
            ApiRouter.sendResponse(exchange, 200, gson.toJson(response));
            
        } catch (Exception e) {
            logger.error("Fehler beim Pairing mit der Hue Bridge {}", bridgeId, e);
            ApiRouter.sendResponse(exchange, 500,
                gson.toJson(Map.of("error", "Fehler beim Pairing mit der Hue Bridge: " + e.getMessage())));
        }
    }
    
    /**
     * Aktualisiert die Bridge im Hue-Modul nach erfolgreichem Pairing.
     * 
     * @param bridgeId Die ID der Bridge
     * @param bridge Das aktualisierte Bridge-Objekt
     */
    private void updateBridgeInModule(String bridgeId, HueDiscoveredBridge bridge) {
        if (moduleRepository == null) {
            logger.debug("ModuleRepository nicht initialisiert - Bridge wird nicht im Modul aktualisiert");
            return;
        }
        
        try {
            String moduleId = "hue";
            Optional<Module> moduleOpt = moduleRepository.findById(moduleId);
            
            if (!moduleOpt.isPresent()) {
                logger.debug("Hue-Modul nicht in Datenbank gefunden - Bridge wird nicht im Modul aktualisiert");
                return;
            }
            
            Module module = moduleOpt.get();
            
            // Initialisiere moduleData falls null
            if (module.getModuleData() == null) {
                module.setModuleData(new java.util.HashMap<>());
            }
            
            // Hole Liste der Bridges aus moduleData und deserialisiere sie korrekt
            List<HueDiscoveredBridge> bridges = new ArrayList<>();
            Object bridgesObj = module.getModuleData().get("bridges");
            
            if (bridgesObj != null) {
                // Deserialisiere die Liste - Gson gibt LinkedTreeMap zurück, daher müssen wir konvertieren
                if (bridgesObj instanceof List) {
                    @SuppressWarnings("unchecked")
                    List<Object> bridgesList = (List<Object>) bridgesObj;
                    for (Object bridgeObj : bridgesList) {
                        if (bridgeObj instanceof HueDiscoveredBridge) {
                            bridges.add((HueDiscoveredBridge) bridgeObj);
                        } else {
                            // Konvertiere LinkedTreeMap zu HueDiscoveredBridge mit Gson
                            String bridgeJson = gson.toJson(bridgeObj);
                            HueDiscoveredBridge deserializedBridge = gson.fromJson(bridgeJson, HueDiscoveredBridge.class);
                            bridges.add(deserializedBridge);
                        }
                    }
                }
            }
            
            // Suche nach der Bridge in der Liste und aktualisiere sie
            boolean found = false;
            for (int i = 0; i < bridges.size(); i++) {
                HueDiscoveredBridge existingBridge = bridges.get(i);
                if (existingBridge != null && bridgeId.equals(existingBridge.getBridgeId())) {
                    // Aktualisiere die Bridge in der Liste
                    bridges.set(i, bridge);
                    found = true;
                    logger.debug("Bridge {} im Hue-Modul aktualisiert", bridgeId);
                    break;
                }
            }
            
            // Falls Bridge nicht in der Liste gefunden wurde, füge sie hinzu
            if (!found) {
                bridges.add(bridge);
                logger.debug("Bridge {} zum Hue-Modul hinzugefügt", bridgeId);
            }
            
            // Speichere aktualisierte Liste im Modul
            module.getModuleData().put("bridges", bridges);
            moduleRepository.save(moduleId, module);
            
            logger.info("Bridge {} erfolgreich im Hue-Modul aktualisiert", bridgeId);
            
        } catch (Exception e) {
            logger.warn("Fehler beim Aktualisieren der Bridge {} im Hue-Modul: {}", bridgeId, e.getMessage());
        }
    }

    /**
     * Erstellt EventStreamManager für alle gepaarten Bridges.
     * Lädt alle Bridges aus der Datenbank und erstellt für jede gepaarte Bridge
     * einen HueModuleEventStreamManager.
     * 
     * @param actionManager Der ActionManager für Device-Zugriff
     * @return Liste der erstellten ModuleEventStreamManager
     */
    public List<ModuleEventStreamManager> createEventStreamManager(ActionManager actionManager) {
        List<ModuleEventStreamManager> managers = new ArrayList<>();
        
        if (bridgeRepository == null || databaseManager == null || actionManager == null) {
            logger.warn("Kann EventStreamManager nicht erstellen: bridgeRepository, databaseManager oder actionManager ist null");
            return managers;
        }
        
        try {
            // Lade alle Bridges aus der Datenbank
            List<HueDiscoveredBridge> bridges = bridgeRepository.findAll();
            
            logger.info("Lade {} Bridges aus der Datenbank für EventStreamManager-Erstellung", bridges.size());
            
            // Erstelle für jede gepaarte Bridge einen EventStreamManager
            for (HueDiscoveredBridge bridge : bridges) {
                if (bridge == null || bridge.getBridgeId() == null) {
                    continue;
                }
                
                // Prüfe, ob die Bridge gepaart ist
                if (bridge.getIsPaired() == null || !bridge.getIsPaired()) {
                    logger.debug("Bridge {} ist nicht gepaart, überspringe EventStreamManager-Erstellung", bridge.getBridgeId());
                    continue;
                }
                
                // Prüfe, ob die Bridge einen Username hat (notwendig für Event-Stream)
                if (bridge.getUsername() == null || bridge.getUsername().isEmpty()) {
                    logger.debug("Bridge {} hat keinen Username, überspringe EventStreamManager-Erstellung", bridge.getBridgeId());
                    continue;
                }
                
                try {
                    // Erstelle EventStreamManager für diese Bridge
                    HueModuleEventStreamManager eventStreamManager = 
                        new HueModuleEventStreamManager(bridge.getBridgeId(), actionManager, databaseManager);
                    managers.add(eventStreamManager);
                    logger.info("EventStreamManager für gepaarte Bridge {} erstellt", bridge.getBridgeId());
                } catch (Exception e) {
                    logger.error("Fehler beim Erstellen des EventStreamManagers für Bridge {}", bridge.getBridgeId(), e);
                }
            }
            
            logger.info("{} EventStreamManager für gepaarte Bridges erstellt", managers.size());
        } catch (Exception e) {
            logger.error("Fehler beim Laden der Bridges aus der Datenbank für EventStreamManager-Erstellung", e);
        }
        
        return managers;
    }
    
    /**
     * Ergebnis-Klasse für den Pairing-Vorgang.
     */
    public static class PairingResult {
        private boolean success = false;
        private boolean error = false;
        private String username;
        private String clientKey;
        private int errorType;
        private String errorDescription;
        private int httpStatus;
        private String rawResponse;
        
        public boolean isSuccess() {
            return success;
        }
        
        public void setSuccess(boolean success) {
            this.success = success;
        }
        
        public boolean isError() {
            return error;
        }
        
        public void setError(boolean error) {
            this.error = error;
        }
        
        public String getUsername() {
            return username;
        }
        
        public void setUsername(String username) {
            this.username = username;
        }
        
        public String getClientKey() {
            return clientKey;
        }
        
        public void setClientKey(String clientKey) {
            this.clientKey = clientKey;
        }
        
        public int getErrorType() {
            return errorType;
        }
        
        public void setErrorType(int errorType) {
            this.errorType = errorType;
        }
        
        public String getErrorDescription() {
            return errorDescription;
        }
        
        public void setErrorDescription(String errorDescription) {
            this.errorDescription = errorDescription;
        }
        
        public int getHttpStatus() {
            return httpStatus;
        }
        
        public void setHttpStatus(int httpStatus) {
            this.httpStatus = httpStatus;
        }
        
        public String getRawResponse() {
            return rawResponse;
        }
        
        public void setRawResponse(String rawResponse) {
            this.rawResponse = rawResponse;
        }
    }
}

File: src/main/node/com/smarthome/backend/server/api/modules/hue/hueBridgeController.ts
import { logger } from "../../../../logger.js";
import https from "node:https";
import type { ClientRequest } from "node:http";
import type { DatabaseManager } from "../../../db/database.js";
import type { HueBridgeDiscovered } from "./hueBridgeDiscovered.js";
import { v3 } from "node-hue-api";
import { HUECONFIG, HUEMODULE } from "./hueModule.js";
import { HueEvent } from "./hueEvent.js";
import { ModuleBridgeControllerEvent } from "../moduleBridgeControllerEvent.js";

/**
 * Initiale Wartezeit (ms) vor einem Reconnect-Versuch nach Verbindungsabbruch.
 */
const INITIAL_RECONNECT_DELAY_MS = 2_000;
/**
 * Maximale Wartezeit (ms) zwischen Reconnect-Versuchen (exponentielles Backoff).
 */
const MAX_RECONNECT_DELAY_MS = 60_000;
/**
 * Falls über diese Zeitspanne (ms) keinerlei Daten/Keepalive von der Bridge
 * empfangen wurden, wird die Verbindung als tot betrachtet und neu aufgebaut.
 * Die Hue Bridge sendet im Eventstream regelmäßig Daten (typ. alle ~10 s),
 * daher ist 90 s ein konservativer Wert.
 */
const STREAM_INACTIVITY_TIMEOUT_MS = 90_000;

interface ActiveEventStream {
  request: ClientRequest;
  inactivityTimer?: NodeJS.Timeout;
  reconnectTimer?: NodeJS.Timeout;
  reconnectAttempts: number;
  stopped: boolean;
  callback: (event: HueEvent) => void;
  bridgeIp: string;
  port: number;
  username: string;
}

export class HueBridgeController extends ModuleBridgeControllerEvent<HueBridgeDiscovered, HueEvent> {

  private activeStreams: Map<string, ActiveEventStream> = new Map();

  constructor(databaseManager: DatabaseManager) {
    super(databaseManager);
  }

  protected getDiscoveredBridgeTypeName(): string {
    return HUECONFIG.bridgeTypeName;
  }
  protected getModuleName(): string {
    return HUEMODULE.id;
  }

  public async pairBridge(bridge: HueBridgeDiscovered): Promise<HueBridgeDiscovered | null> {
    if (!bridge) return null;
    const bridgeId = bridge.id;
    if ( !bridgeId ) return null;
    const ipAddress = bridge.address;
    if (!ipAddress) return null;

    try {
      const unauthenticatedApi = await v3.api.createLocal(ipAddress).connect();
      const createdUser = await unauthenticatedApi.users.createUser("smarthome-backend", "server");
      bridge.isPaired = true;
      bridge.username = createdUser.username;
      bridge.clientKey = createdUser.clientkey;
      super.saveBridge(bridgeId, bridge);
      return bridge;
    } catch (err: unknown) {
      logger.warn({ err, bridgeId }, "Hue Pairing fehlgeschlagen für {} ({})", bridgeId, err instanceof Error ? err.message : String(err));
      return null;
    }
  }

  public async startEventStream(bridge: HueBridgeDiscovered, callback: (event: HueEvent) => void): Promise<void> {
    if( bridge.isPaired && bridge.username && bridge.address ) {
      const port = this.resolveEventStreamPort(bridge.port);

      const existing = this.activeStreams.get(bridge.id);
      if (existing) {
        if (!existing.stopped) {
          logger.warn({ bridgeId: bridge.id }, "Hue EventStream laeuft bereits, starte nicht erneut");
          return;
        }
        this.cleanupStream(bridge.id);
      }

      const stream: ActiveEventStream = {
        request: undefined as unknown as ClientRequest,
        reconnectAttempts: 0,
        stopped: false,
        callback,
        bridgeIp: bridge.address,
        port,
        username: bridge.username
      };
      this.activeStreams.set(bridge.id, stream);
      this.openEventStreamConnection(bridge.id);
    } else {
      logger.error({ bridgeId: bridge.id }, "Hue EventStream kann nicht gestartet werden: Bridge ist nicht gepaart oder hat keine IP-Adresse");
      throw new Error("Hue EventStream kann nicht gestartet werden: Bridge ist nicht gepaart oder hat keine IP-Adresse");
    }
  }

  public async stopEventStream(bridge: HueBridgeDiscovered): Promise<void> {
    const stream = this.activeStreams.get(bridge.id);
    if (!stream) {
      logger.debug({ bridgeId: bridge.id }, "Hue EventStream stop: keine aktive Verbindung");
      return;
    }
    stream.stopped = true;
    this.cleanupStream(bridge.id);
    logger.info({ bridgeId: bridge.id }, "Hue EventStream gestoppt");
  }

  private cleanupStream(bridgeId: string) {
    const stream = this.activeStreams.get(bridgeId);
    if (!stream) return;
    if (stream.inactivityTimer) {
      clearTimeout(stream.inactivityTimer);
      stream.inactivityTimer = undefined;
    }
    if (stream.reconnectTimer) {
      clearTimeout(stream.reconnectTimer);
      stream.reconnectTimer = undefined;
    }
    try {
      stream.request?.destroy();
    } catch (err) {
      logger.debug({ err, bridgeId }, "Fehler beim Beenden des Hue Eventstream-Requests");
    }
    if (stream.stopped) {
      this.activeStreams.delete(bridgeId);
    }
  }

  /**
   * Plant einen Reconnect-Versuch mit exponentiellem Backoff.
   * Wird kein Reconnect-Versuch geplant, falls der Stream als gestoppt markiert ist.
   */
  private scheduleReconnect(bridgeId: string, reason: string): void {
    const stream = this.activeStreams.get(bridgeId);
    if (!stream || stream.stopped) return;
    if (stream.reconnectTimer) return; // bereits geplant

    const attempt = stream.reconnectAttempts;
    const delay = Math.min(INITIAL_RECONNECT_DELAY_MS * Math.pow(2, attempt), MAX_RECONNECT_DELAY_MS);
    stream.reconnectAttempts = attempt + 1;

    logger.warn(
      { bridgeId, bridgeIp: stream.bridgeIp, reason, delayMs: delay, attempt: stream.reconnectAttempts },
      "Hue EventStream Reconnect geplant"
    );

    stream.reconnectTimer = setTimeout(() => {
      const current = this.activeStreams.get(bridgeId);
      if (!current || current.stopped) return;
      current.reconnectTimer = undefined;
      this.openEventStreamConnection(bridgeId);
    }, delay);
  }

  private armInactivityTimer(bridgeId: string): void {
    const stream = this.activeStreams.get(bridgeId);
    if (!stream || stream.stopped) return;
    if (stream.inactivityTimer) clearTimeout(stream.inactivityTimer);
    stream.inactivityTimer = setTimeout(() => {
      logger.warn(
        { bridgeId, bridgeIp: stream.bridgeIp, timeoutMs: STREAM_INACTIVITY_TIMEOUT_MS },
        "Hue EventStream Inaktivitaets-Timeout, erzwinge Reconnect"
      );
      try {
        stream.request?.destroy();
      } catch (err) {
        logger.debug({ err, bridgeId }, "Fehler beim Schliessen des inaktiven Hue Eventstream-Requests");
      }
      // 'close' bzw. 'error' loest dann scheduleReconnect aus
    }, STREAM_INACTIVITY_TIMEOUT_MS);
  }

  private openEventStreamConnection(bridgeId: string): void {
    const stream = this.activeStreams.get(bridgeId);
    if (!stream || stream.stopped) return;

    const { bridgeIp, port, username, callback } = stream;
    const eventStreamUrl = `https://${bridgeIp}:${port}/eventstream/clip/v2`;
    const agent = new https.Agent({ rejectUnauthorized: false, keepAlive: true });

    const request = https.request(
      eventStreamUrl,
      {
        method: "GET",
        headers: {
          "hue-application-key": username,
          Accept: "text/event-stream"
        },
        agent
      },
      response => {
        const current = this.activeStreams.get(bridgeId);
        if (!current || current.stopped) {
          response.resume();
          return;
        }

        if (response.statusCode !== 200) {
          logger.error(
            { bridgeId, bridgeIp, statusCode: response.statusCode },
            "Hue Eventstream-Verbindung fehlgeschlagen"
          );
          response.resume();
          this.scheduleReconnect(bridgeId, `statusCode=${response.statusCode}`);
          return;
        }

        logger.info({ bridgeId, bridgeIp }, "Hue EventStream verbunden");
        current.reconnectAttempts = 0;
        this.armInactivityTimer(bridgeId);

        response.setEncoding("utf8");
        let eventData = "";
        let buffer = "";

        response.on("data", (chunk: string) => {
          this.armInactivityTimer(bridgeId);

          buffer += chunk;
          let index = buffer.indexOf("\n");
          while (index !== -1) {
            const line = buffer.slice(0, index).replace(/\r$/, "");
            buffer = buffer.slice(index + 1);
            if (line.startsWith("data:")) {
              const data = line.slice(5).trim();
              eventData += data;
              if (data.endsWith("]")) {
                this.processEventData(bridgeId, eventData, callback);
                eventData = "";
              }
            } else if (line.length === 0) {
              if (eventData.length > 0) {
                this.processEventData(bridgeId, eventData, callback);
                eventData = "";
              }
            }
            index = buffer.indexOf("\n");
          }
        });

        response.on("end", () => {
          logger.info({ bridgeId, bridgeIp }, "Hue EventStream-Verbindung beendet");
          this.scheduleReconnect(bridgeId, "response-end");
        });

        response.on("error", err => {
          logger.warn({ err, bridgeId, bridgeIp }, "Hue EventStream Response-Fehler");
          this.scheduleReconnect(bridgeId, "response-error");
        });
      }
    );

    stream.request = request;

    // TCP keep-alive aktivieren, damit das Betriebssystem tote Verbindungen
    // (z. B. nach NAT-Timeouts oder Bridge-Reboot) erkennen kann.
    request.on("socket", socket => {
      socket.setKeepAlive(true, 30_000);
    });

    request.on("error", err => {
      logger.warn({ err, bridgeId, bridgeIp }, "Fehler im Hue Eventstream-Request");
      this.scheduleReconnect(bridgeId, "request-error");
    });

    request.on("close", () => {
      const current = this.activeStreams.get(bridgeId);
      if (!current || current.stopped) return;
      logger.debug({ bridgeId, bridgeIp }, "Hue EventStream-Request geschlossen");
      this.scheduleReconnect(bridgeId, "request-close");
    });

    request.end();
  }

  private processEventData(bridgeId: string, eventDataJson: string, callback: (event: HueEvent) => void) {
    try {
      const events = JSON.parse(eventDataJson) as Array<Record<string, unknown>>;
      for (const event of events ?? []) {
        const dataArray = (event as any).data as Array<Record<string, unknown>> | undefined;
        if (!Array.isArray(dataArray)) continue;
        for (const data of dataArray) {
          callback({
            bridgeId: bridgeId,
            data: data
          });
        }
      }
    } catch (err) {
      logger.warn({ err, bridgeId: bridgeId }, "Fehler beim Parsen von Event-Daten");
    }
  }

  private resolveEventStreamPort(port?: number) {
    if (!port || port === 80) return 443;
    return port;
  }
}
